### ISSUE TITLE
[Bug]: LoRA lm_head delta is silently dropped from prompt_logprobs

### ISSUE THREAD

**Issue by XiaohanZhangCMU** [vllm-project/vllm#51594]
### Summary

When a LoRA adapter targets `lm_head`, its delta is **silently dropped from `prompt_logprobs`**. Generation logprobs are correct; teacher-forced prompt logprobs for the same tokens are not. Nothing warns, and the returned numbers look plausible.

### Evidence

Measured on 2x B300, TP=2, vLLM `0.26.1rc1.dev245+ge2fa28594`, a small model deliberately overfit on 8 fixed examples so the adapter carries a large, known delta. The trainer that produced the adapter reports mean logprob **-0.0014 nats/token** on 991 supervised positions.

Same 991 positions, same token ids, one served adapter:

| reading | mean logprob |
|---|---|
| training framework (reference) | -0.0014 |
| vLLM, generation logprobs | **-0.0091** |
| vLLM, `prompt_logprobs` | **-6.63** |
| vLLM, `prompt_logprobs`, no adapter | -12.12 |

Greedy decoding reproduces the memorized targets verbatim 8/8, so the adapter is loaded and working. `prompt_logprobs` lands between "adapter applied" and "no adapter".

### It is specifically the lm_head LoRA

Taking the same adapter and deleting only its two lm_head tensors from the safetensors:

| adapter | generation | `prompt_logprobs` | gap |
|---|---|---|---|
| full, incl. lm_head | -0.0091 | **-6.6265** | 6.6174 |
| lm_head tensors removed | -6.7497 | **-6.6265** | 0.1232 |

`prompt_logprobs` returns the same value to four decimals whether the lm_head LoRA is present or absent. It contributes nothing on that path. Attention LoRA is unaffected, because it is already in `hidden_states` before logits are computed.

### Root cause

`vllm/v1/worker/gpu_model_runner.py`, `_get_prompt_logprobs_dict`:

```python
prompt_hidden_states = hidden_states[offset : offset + num_logits]
logits = self.model.compute_logits(prompt_hidden_states)
```

`compute_logits` reaches `LogitsProcessorWithLoRA._get_logits`, which applies the delta with:

```python
# vllm/lora/punica_wrapper/punica_gpu.py, add_lora_logits
lora_shrink(x, [lora_a_stacked], buffer.unsqueeze(0),
            *self.prompt_mapping_meta.meta_args(x.size(0), ...), scale)
```

`prompt_mapping_meta` is built from the **sampler** mapping, one entry per sequence:

```python
# punica_gpu.py, _update_prefill_metadata
self.token_mapping_meta.prepare_tensors(self.token_lora_indices)   # per token
self.prompt_mapping_meta.prepare_tensors(self.sampler_indices)     # per sequence
```

`add_lora_logits` is documented as "specifically for LogitsProcessorWithLoRA", and that assumption holds for sampling, where there is exactly one logits row per sequence. The prompt-logprobs path breaks it: it passes `num_logits` rows, one per prompt **token**, so the mapping does not describe those rows and the lm_head delta is not applied to them.

### Reproduce

1. Serve any model with `--enable-lora` and an adapter whose `target_modules` includes `lm_head` (requires untied embeddings).
2. Pick a sequence the adapter strongly changes.
3. Compare, for the same token ids: `POST /v1/completions` with `prompt_logprobs: 0`, against the per-token `logprobs` returned when generating those tokens.
4. They should agree. They do not, and stripping the lm_head tensors from the adapter makes `prompt_logprobs` unchanged.

### Impact

Non-fatal and silent, which is what makes it costly. It affects anyone scoring text with `prompt_logprobs` against a LoRA deployment: RL training loops that compute importance ratios or KL from teacher-forced scores, offline eval, and reward models. Generation is unaffected, so a system can look healthy while its scoring is systematically wrong. In our case the drift is 6.6 nats/token.

### Attempted fix that did not work

Recording this so it is not retried blindly. Since the correct per-token mapping already exists as `token_lora_indices`, we tried borrowing the matching slice around `compute_logits`:

```python
wrapper.prompt_mapping_meta.prepare_tensors(wrapper.token_lora_indices[start:end])
try:
    logits = self.model.compute_logits(prompt_hidden_states[s:e])
finally:
    wrapper.prompt_mapping_meta.prepare_tensors(wrapper.sampler_indices)
```

Measured after the change: unchanged, -6.6265. We have not yet established whether the swap did not take effect (our lookup of the language-model punica wrapper may return `None` on a multimodal model, in which case it silently no-oped) or whether re-preparing that metadata is not sufficient. We did not want to open a PR around a change we cannot show working. Happy to test any candidate patch on this setup and report numbers.

### Environment

vLLM `0.26.1rc1.dev245+ge2fa28594` (nightly, cu130), torch 2.13.0+cu130, 2x B300, TP=2, expert parallel, V1 engine, `max_loras=2`, `max_lora_rank=32`, untied embeddings. The same code shape is present on current `main`.

**Comment by jialoop-git** [comment]
@XiaohanZhangCMU I've opened a fix in #51597. The key change is temporarily swapping prompt_mapping_meta to per-token LoRA indices before each compute_logits call in _get_prompt_logprobs_dict. Would you be able to run the numbers on your B300 setup?
